plugins/modules/user.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2019, Paul Arthur <paul.arthur@flowerysong.com>
+# Copyright: (c) 2019, XLAB Steampunk <steampunk@xlab.si>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'XLAB Steampunk'}
+
+DOCUMENTATION = '''
+module: user
+author:
+  - Paul Arthur (@flowerysong)
+  - Aljaz Kosir (@aljazkosir)
+  - Tadej Borovsak (@tadeboro)
+short_description: Manages Sensu users
+description:
+  - For more information, refer to the Sensu documentation at
+    U(https://docs.sensu.io/sensu-go/latest/reference/users/)
+extends_documentation_fragment:
+  - sensu.sensu_go.base
+notes:
+  - Parameter C(auth.namespace) is ignored in this module.
+options:
+  state:
+    description:
+      - Desired state of the user.
+      - Users cannot actually be deleted, only deactivated.
+    type: str
+    choices: ['enabled', 'disabled']
+    default: enabled
+  name:
+    description:
+      - Username.
+    type: str
+    required: true
+  password:
+    description:
+      - Password for the user.
+    type: str
+  groups:
+    description:
+      - List of groups user belongs to.
+    type: list
+'''
+
+EXAMPLES = '''
+- name: Create a user
+  user:
+    auth:
+      url: http://localhost:8080
+    name: awesome_username
+    password: hidden_password?
+    groups:
+      - dev
+      - prod
+'''
+
+RETURN = '''
+object:
+    description: object representing Sensu user
+    returned: success
+    type: dict
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+
+from ansible_collections.sensu.sensu_go.plugins.module_utils import (
+    arguments, errors, utils,
+)
+
+
+def do_differ(current, desired):
+    if current is None:
+        return True
+
+    for key, value in desired.items():
+        if key == 'password':
+            return True  # We can't know if the password changed or not
+        if value != current.get(key):
+            return True
+
+    return False
+
+
+def sync(remote_object, state, client, path, payload, check_mode):
+    if state == 'disabled' and remote_object is not None:
+        if not check_mode:
+            utils.delete(client, path)
+        return True, utils.get(client, path)
+
+    if do_differ(remote_object, payload):
+        if check_mode:
+            return True, payload
+        utils.put(client, path, payload)
+        return True, utils.get(client, path)
+
+    return False, remote_object
+
+
+def is_password_required(remote_object, state):
+    return not (state == 'disabled' and remote_object is not None)
```

Comment:
If we apply the de Morgan's law here, we can convert the condition into

    state != "disabled" or remote_object is None

Both parts of this condition are a bit counter-intuitive. I guess we cannot do much about the first part since this condition seems to be directly linked to the API requirements. But maybe we can do something about the second part?

I am a bit worried that our end-users will find it a bit counter-intuitive if they need to specify a password when disabling a non-existing backend user. Maybe we should create a disabled user only if the `password` parameter has been set? If we decide to go this way, having a task

    sensu.sensu_go.user:
      name: Lizzy
      state: disabled

in the playbook would not create new user `Lizzy` if she does not exist yet. On the other hand, putting a task

    sensu.sensu_go.user:
      name: Lizzy
      state: disabled
      password: pass

would create a new (disabled) user Lizzy with the password set to `pass`. Any opinions about this approach?

Reply:
I like it, and we can remove the `is_password_required` entirely and only use ansible's `required_if`. If user tries to disable a non existent user we stop the module and tell him that he cannot do that.